// @uniswap/universal-router/typechain/factories/V3Path__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Signer } from "ethers";
import { Provider, TransactionRequest } from "@ethersproject/providers";
import { Contract, ContractFactory, Overrides } from "@ethersproject/contracts";

import type { V3Path } from "../V3Path";

export class V3Path__factory extends ContractFactory {
  constructor(signer?: Signer) {
    super(_abi, _bytecode, signer);
  }

  deploy(overrides?: Overrides): Promise<V3Path> {
    return super.deploy(overrides || {}) as Promise<V3Path>;
  }
  getDeployTransaction(overrides?: Overrides): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  attach(address: string): V3Path {
    return super.attach(address) as V3Path;
  }
  connect(signer: Signer): V3Path__factory {
    return super.connect(signer) as V3Path__factory;
  }
  static connect(address: string, signerOrProvider: Signer | Provider): V3Path {
    return new Contract(address, _abi, signerOrProvider) as V3Path;
  }
}

const _abi = [
  {
    inputs: [],
    name: "SliceOutOfBounds",
    type: "error",
  },
];

const _bytecode =
  "0x6080806040523460175760119081601d823930815050f35b600080fdfe600080fdfea164736f6c6343000811000a";
